crates/sui-graphql-client/src/error.rs

```diff
@@ -0,0 +1,316 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::num::ParseIntError;
+use std::num::TryFromIntError;
+
+use cynic::GraphQlError;
+
+use sui_types::types::AddressParseError;
+use sui_types::types::DigestParseError;
+use sui_types::types::TypeParseError;
+
+type BoxError = Box<dyn std::error::Error + Send + Sync>;
+
+/// General error type for the client. It is used to wrap all the possible errors that can occur.
+#[derive(Debug)]
+pub struct Error {
+    inner: Box<ClientError>,
+}
+
+/// Error type for the client. It is split into multiple fields to allow for more granular error
+/// handling. The `source` field is used to store the original error.
+#[derive(Debug)]
+pub struct ClientError {
+    /// Error when the GraphQL server returns an error.
+    pub query_response_error: bool,
+    /// Error when deserialization a value (mostly bcs or base64).
+    pub deserialize_error: bool,
+    /// Error when parsing a value.
+    pub parse_error: bool,
+    /// Graphql server returned an empty response, though it was expected to return a value.
+    pub empty_response_error: bool,
+    /// Error when converting from a type to another.
+    pub conversion_error: bool,
+    /// The original error. Use downcasting to get the original error based on the error type from
+    /// above.
+    pub source: Option<BoxError>,
+}
+
+/// An empty response with no data from the GraphQL server. This is used to signal that the API
+/// expected some data, but the response was empty.
+#[derive(Debug, Clone)]
+pub struct EmptyResponse;
+
+/// Error when deserialization a value (mostly bcs or base64).
+#[derive(Debug, Clone)]
+pub struct DeserializeError(pub String);
+
+/// Error when parsing a value.
+#[derive(Debug, Clone)]
+pub enum ParseError {
+    /// Error when parsing a URL.
+    UrlParseError(String),
+    /// Error when parsing a BigInt into integer.
+    BigIntParsingError,
+}
+
+/// Error when converting from a type to another.
+#[derive(Debug, Clone)]
+pub struct ConversionError(pub String);
+
+/// Error when the GraphQL server returns an error.
+#[derive(Debug, Clone)]
+pub struct QueryResponseError(pub Vec<GraphQlError>);
+
+#[derive(Debug, Clone)]
+pub struct ConflictingArguments(pub String);
```

Comment:
I don't think we need each of these error types. Instead we can have an enum `Kind` that can indicate what type of error it is. All source error information should be stored in `source` in its original, un-stringified form. We can either choose to expose the kind and make it public (this would be easiest) so long as its marked as "non-exhaustive" or have "is_XXX" functions for doing those checks ourselves. I think making kind public is probably the easiest thing to do from a maintainability point of view

Reply:
For the list of GraphQLErrors, since we are tightly coupled to cynic, we can had an Option<Vec<GraphQlError>> field to the Inner Error that can be exposed via a public function `pub fn graphql_errors(&self) -> Option<&[GraphQLError]>`